### PR TITLE
feat: use env var to specify npm path

### LINKED ISSUE #1790: Config option to specify npm location for `pnpm run`
Pnpm invokes npm for commands like `pnpm run` and silently fails if npm is not in $PATH:

https://github.com/pnpm/pnpm/blob/8132523244fbc1765d13fed7a56959616c974458/packages/pnpm/src/cmd/runNpm.ts#L4

I remove npm from $PATH in my development containers to avoid calling `npm install` accidentally.

Would it be possible to add a configuration option to specify npm location, e.g., `/opt/npm/bin/npm`?

P.S. I have switched to pnpm after suffering through numerous cache issues with yarn. Thanks for a great package manager!

### PR DESCRIPTION
**Pull request by yaodingyd**
Close #1790 

I think using a env var makes more sense but happy to take feedback.

### CODE REVIEW

**Comment by zkochan** [comment]
The config settings are also read from the env. So if `npm-path` is added here https://github.com/pnpm/pnpm/blob/master/packages/config/src/index.ts#L84

it'll be possible to also set it via `NPM_CONFIG_NPM_PATH`. We reuse npm's config lib that's why tit should be prefixed with NPM_CONFIG

**Comment by yaodingyd** [comment]
After we add this config entry here, what do you prefer for the next step?

- add `@pnpm/config` as a dependency for `@pnpm/run-npm` so we can just do this config within `runNpm` function, or
- make `runNpm` to take another argument `npmPath`, and update all its usages?

**Comment by zkochan** [comment]
> make runNpm to take another argument npmPath, and update all its usages?

this. You'll have to add also a new type to the Config type in `@pnpm/config` and then any plugin that uses `@pnpm/run-npm` will have to pick that type in its options.

File: packages/plugin-commands-publishing/src/runNpm.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    cliOptions: {},
+    packageManager,
+    rcOptionsTypes: {
+      ...R.pick([
+        'npm-path',
+      ], allTypes),
+    },
+  })
+  return runNpm(config.npmPath, args)
+}
```
Comment (zkochan) [inline packages/plugin-commands-publishing/src/runNpm.ts]: This file is not needed. The configs are read on each command execution, so you just repeat what already was done.

You should update the following line to contain this npm-path:

https://github.com/pnpm/pnpm/blob/a56c7856a2ead878bc4dc4285514da9b1e84175c/packages/plugin-commands-publishing/src/publish.ts#L17

So something like

```ts
export function rcOptionsTypes () {
return {
  ...cliOptionsTypes(),
  ...R.pick([
      'npm-path',
    ], allTypes)
}
```

And here you should add `'npm-path' |`:

https://github.com/pnpm/pnpm/blob/a56c7856a2ead878bc4dc4285514da9b1e84175c/packages/plugin-commands-publishing/src/recursivePublish.ts#L31

That way, `npmPath` will be in the opts passed to the publishing plugin.

Same changes should be done in other plugins that run npm

File: packages/pnpm/src/runNpm.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    cliOptions: {},
+    packageManager,
+    rcOptionsTypes: {
+      ...R.pick([
+        'npm-path',
+      ], allTypes),
+    },
+  })
+  return runNpm(config.npmPath, args)
+}
```
Comment (zkochan) [inline packages/pnpm/src/runNpm.ts]: This file is not used, so should be removed

**Comment by zkochan** [comment]
Seems correct.

The tests are failing in `plugin-commands-publishing`. Maybe the new config setting should be added there as well

**Comment by yaodingyd** [comment]
Dumb question: when I run `pnpm test` in workspace root, I got `Missing script: test` and I do notice test is missing in root level package.json. In which step I did wrong?

**Comment by zkochan** [comment]
you can run `pnpm run test-master` to run all test
or `pnpm run test-branch` to test changed packages only

**Pull request merged**